junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/EngineDiscoveryResultValidator.java

```diff
@@ -37,30 +38,39 @@ void validate(TestEngine testEngine, TestDescriptor root) {
 			() -> String.format(
 				"The discover() method for TestEngine with ID '%s' must return a non-null root TestDescriptor.",
 				testEngine.getId()));
-		Preconditions.condition(isAcyclic(root),
-			() -> String.format("The discover() method for TestEngine with ID '%s' returned a cyclic graph.",
-				testEngine.getId()));
+		Optional<UniqueId> cyclicUIDs = getFirstCyclicUID(root);
+		Preconditions.condition(!cyclicUIDs.isPresent(),
+			() -> String.format("The discover() method for TestEngine with ID '%s' returned a cyclic graph: "
+					+ "Cycle found at unique ID '%s'",
```

Comment:
I think it would be even more helpful to print all unique IDs in the cycle or at least the two between which the cycle was detected (`detectInnerClassCycle` in `ReflectionUtils` could be used for inspiration).

@tribbloid Would you have time to look into that?

Reply:
Absolutely, that's why I opted to use List instead of Optional in the first place.

I just don't know if it can halt every time, should we add a some kind of depth check?

Reply:
Here's an article outlining how to find the shortest cycle in an undirected graph (tests should be a tree, i.e. an acyclic undirected graph): https://www.baeldung.com/cs/shortest-cycle-undirected-graph

Reply:
@marcphilipp thanks a lot, I'll add the code to print the first detected path of the cycle (using `node.parent()`).

But I'm reluctant to change the algorithm from breadth first (the status quo) to depth first (one used in your link), the search will hang for diverging tree without loop (e.g. if you generate a UUID suffix for each new sub-test). Existing code probably chose breadth first for this reason (and some other reasons)

Reply:
OK the latest  version now print 2 paths that leads to the same node. There are some strange styling errors which shouldn't affect implementation.

All 3 test cases now contains examples of the error messages

Reply:
LGTM. Let me know if it looks good or only requires minor revision.

Reply:
> Now the styling error is also fixed. Only get some deprecation error under JVM 20 that is irrelevant to the scope of this change.

Can we close it now?